File: packages/typescript/src/judges/weightedVote.ts
import type { DebateTranscript } from "../debate/engine.ts";
import { Verdict, fallbackVerdict } from "./base.ts";
import type { JudgeStrategy } from "./base.ts";

export class WeightedVoteJudge implements JudgeStrategy {
  async judge(transcript: DebateTranscript, _labels: string[]): Promise<Verdict> {
    const finalRound = transcript.rounds[transcript.rounds.length - 1] ?? [];

    if (finalRound.length === 0) {
      return fallbackVerdict(transcript, "weighted_vote");
    }

    const scores = new Map<string, number>();
    for (const response of finalRound) {
      scores.set(response.label, (scores.get(response.label) ?? 0) + Number(response.confidence));
    }

    let winner = "";
    let bestScore = -1;
    for (const [label, score] of scores.entries()) {
      if (score > bestScore) {
        winner = label;
        bestScore = score;
      }
    }

    const total = Array.from(scores.values()).reduce((a, b) => a + b, 0) || 1;
    return new Verdict({
      label: winner,
      confidence: bestScore / total,
      reasoning: "Weighted vote based on persona confidence scores.",
      wasEscalated: true,
      primaryResult: transcript.primaryResult,
      debateTranscript: transcript,
      judgeStrategy: "weighted_vote",
      totalDurationMs: transcript.durationMs,
      totalCostUsd: transcript.totalCostUsd,
    });
  }
}

File: packages/python/src/llm_jury/judges/weighted_vote.py
from __future__ import annotations

from collections import defaultdict

from llm_jury.debate.engine import DebateTranscript

from .base import JudgeStrategy, Verdict, _fallback_verdict


class WeightedVoteJudge(JudgeStrategy):
    async def judge(self, transcript: DebateTranscript, labels: list[str]) -> Verdict:
        if not transcript.rounds or not transcript.rounds[-1]:
            return _fallback_verdict(transcript, "weighted_vote")

        final_round = transcript.rounds[-1]
        scores: dict[str, float] = defaultdict(float)

        for response in final_round:
            scores[response.label] += float(response.confidence)

        winner = max(scores, key=lambda k: scores[k])
        total = sum(scores.values()) or 1.0
        confidence = float(scores[winner] / total)

        return Verdict(
            label=winner,
            confidence=confidence,
            reasoning="Weighted vote based on persona confidence scores.",
            was_escalated=True,
            primary_result=transcript.primary_result,
            debate_transcript=transcript,
            judge_strategy="weighted_vote",
            total_duration_ms=transcript.duration_ms,
            total_cost_usd=transcript.total_cost_usd,
        )
